contrib/jinzhu/gorm/gorm.go

```diff
@@ -7,17 +7,109 @@
 package gorm
 
 import (
+	"context"
+	"math"
+	"time"
+
 	sqltraced "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 
 	"github.com/jinzhu/gorm"
 )
 
+const (
+	gormContextKey       = "dd-trace-go:context"
+	gormConfigKey        = "dd-trace-go:config"
+	gormSpanStartTimeKey = "dd-trace-go:span"
+)
+
 // Open opens a new (traced) database connection. The used dialect must be formerly registered
 // using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
-func Open(dialect, source string) (*gorm.DB, error) {
-	db, err := sqltraced.Open(dialect, source)
+func Open(dialect, source string, opts ...Option) (*gorm.DB, error) {
+	sqldb, err := sqltraced.Open(dialect, source)
 	if err != nil {
 		return nil, err
 	}
-	return gorm.Open(dialect, db)
+	db, err := gorm.Open(dialect, sqldb)
+	if err != nil {
+		return db, err
+	}
+	return WithCallbacks(db, opts...), err
+}
+
+// WithCallbacks registers callbacks to the gorm.DB for tracing.
+func WithCallbacks(db *gorm.DB, opts ...Option) *gorm.DB {
+	cb := db.Callback()
+	cb.Create().Before("gorm:before_create").Register("dd-trace-go:before_create", before)
+	cb.Create().After("gorm:after_create").Register("dd-trace-go:after_create", afterFunc("gorm.create"))
+	cb.Update().Before("gorm:before_update").Register("dd-trace-go:before_update", before)
+	cb.Update().After("gorm:after_update").Register("dd-trace-go:after_update", afterFunc("gorm.update"))
+	cb.Delete().Before("gorm:before_delete").Register("dd-trace-go:before_delete", before)
+	cb.Delete().After("gorm:after_delete").Register("dd-trace-go:after_delete", afterFunc("gorm.delete"))
+	cb.Query().Before("gorm:query").Register("dd-trace-go:before_query", before)
+	cb.Query().After("gorm:after_query").Register("dd-trace-go:after_query", afterFunc("gorm.query"))
+	cb.RowQuery().Before("gorm:row_query").Register("dd-trace-go:before_row_query", before)
+	cb.RowQuery().After("gorm:row_query").Register("dd-trace-go:after_row_query", afterFunc("gorm.row_query"))
+
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return db.Set(gormConfigKey, cfg)
+}
+
+// WithContext returns a new gorm.DB with the context added
+// to its settings store.
+func WithContext(ctx context.Context, db *gorm.DB) *gorm.DB {
+	if ctx == nil {
+		return db
+	}
+	db = db.Set(gormContextKey, ctx)
+	return db
+}
+
+func before(scope *gorm.Scope) {
```

Comment:
Can we inline this function and the `afterFunc` function inside `WithCallbacks`? It's not used anywhere and it's short enough to make it clear what happens. I means as variable declaration:
```go
beforeFunc := func(scope *gorm.Scope) { scope.Set(gormSpanStartTimeKey, time.Now()) }
afterFunc := ...
```

Reply:
Sure. I don't have strong feelings about it.